### PR TITLE
Impl trait diagnostic tweaks

### LINKED ISSUE #119629: Error messages for where `impl Trait` is allowed do not account for return-position-impl-trait-in-trait (RPITIT)
### Code

```Rust
fn repro(a: impl Fn() -> impl Drop) {}
```

### Current output

```
error[E0562]: `impl Trait` only allowed in function and inherent method argument and return types, not in `Fn` trait return types
 --> src/lib.rs:1:26
  |
1 | fn repro(a: impl Fn() -> impl Drop) {}
  |                          ^^^^^^^^^
```

### Desired output

I believe this message should mention the newly-stabilized possibilities of RPITIT. For example, this is not a function or an inherent method:

```rust
trait Moo {
    fn moo() -> impl Drop;
}
```

### Rationale and extra context

_No response_

### Other cases

My gut tells me that there are other errors with the same now-incorrect subset of locations.

### Anything else?

Seen in nightly (2024-01-04 f688dd684faca5b31b15)

### PR DESCRIPTION
**Pull request by compiler-errors**
1. Tweak some names for `impl Trait` being used in the wrong position
2. Remove two helper functions that are no longer needed since RPITIT is stable, and which causes matches to be a bit obtuse.
3. Split and fix the part where the error notes that it's "only allowed in XX"

Fixes #119629

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? fmease or feel free to reassign if you don't want to review it 😆

File: tests/ui/impl-trait/where-allowed.rs [hunk 1]
```diff
@@ -155,97 +155,97 @@ type InTypeAlias<R> = impl Debug;
 //~^ ERROR `impl Trait` in type aliases is unstable
 
 type InReturnInTypeAlias<R> = fn() -> impl Debug;
-//~^ ERROR `impl Trait` only allowed in function and inherent method argument and return types
+//~^ ERROR `impl Trait` is not allowed in `fn` pointer
```
Comment (fmease) [inline tests/ui/impl-trait/where-allowed.rs:158]: `` `fn` pointers `` plural sounds better I guess. Odd that we don't say `` `fn` pointer return types `` here while we do say that in other places

File: tests/ui/impl-trait/issues/issue-79099.rs [hunk 2]
```diff
@@ -1,7 +1,7 @@
 struct Bug {
     V1: [(); {
         let f: impl core::future::Future<Output = u8> = async { 1 };
-        //~^ `impl Trait` only allowed in function and inherent method argument and return types
+        //~^ `impl Trait` is not allowed in variable bindings
```
Comment (fmease) [inline tests/ui/impl-trait/issues/issue-79099.rs:4]: `in variable bindings` is fine but `in the type of variable bindings` imo, not super important

**Review by fmease — APPROVED** [review]
Thanks, very nice! r=me with or without nits fixed

File: tests/ui/impl-trait/where-allowed.rs (hunk 1 above)
Comment (compiler-errors) [reply tests/ui/impl-trait/where-allowed.rs:158]: sorry lol we do say the whole thing, i just only matched a subset of the error message

**Comment by compiler-errors** [comment]
wait i'll fix that variable binding nit

**Comment by fmease** [comment]
ok ok :)

**Pull request merged**